glambie/config/yaml_helpers.py

```diff
@@ -0,0 +1,22 @@
+
+def strip_python_tags(yaml_string: str):
+    """
+    Strips the python tags from a yaml
+
+    Parameters
+    ----------
+    yaml_string : str
+        yaml string
+
+    Returns
+    -------
+    _type_
+        stripped yaml string
+    """
+    result = []
+    for line in yaml_string.splitlines():
+        idx = line.find("!!python/")
```

Comment:
This suggests that you are trying to dump something to a YAML that is not a basic python type, so the problem you are trying to solve with this isn't actually here, and shouldn't be solved this way.

If the thing you're saving to yaml only contains basic types, the string `!!python/` will never occur in that yaml. My guess is that this has something to do with sets or perhaps datetimes? At minimum you need to explain what type you are trying to suppress/clean here in the docstring.

Reply:
Yes, so I am dumping a config object (RegionRunConfig), so the !!python/ shows up in the first line, and then also for the "year_type" which is an Enum.

In RegionRunConfig: 
```
    def save_to_yaml(self, out_path):
        yaml_string = yaml.dump(self)
        yaml_string = strip_python_tags(yaml_string)
        with open(out_path, 'w') as outfile:
            outfile.write(yaml_string)
```

What would be a less hacky way to solve this?

Reply:
Given that you already have a save_to_yaml defined for RegionRunConfig, and you're not interested in loading the run config again (at least not at present), the simplest way is probably adapting something like:

```python
import yaml

class MyAClass:
    def __init__(self, a) -> None:
        self.a = a

class MyBClass:
    def __init__(self, a) -> None:
        self.a = a        


def main():
    # create a list of custom classes, some of which contain nested classes
    stuff = [MyAClass(1), MyAClass(MyBClass(2)), MyAClass(MyBClass(MyAClass(3)))]

    # directly dumping a list of MyClass produces valid but somewhat ugly output, if the intent is human consumption
    print('this is pretty ugly:')
    print(yaml.dump(stuff))

    # instead, we can use a custom representer to forcibly convert the object to its own dictionary,
    # which generally looks nicer:
    def my_class_representer(yaml_dumper, object_to_represent):
        return yaml_dumper.represent_dict(object_to_represent.__dict__)

    yaml.add_representer(MyAClass, my_class_representer)
    yaml.add_representer(MyBClass, my_class_representer)    

    print('this is nicer:')
    print(yaml.dump(stuff))


if __name__ == '__main__':
    main()
```

Reply:
So this is nice and neat, and I tested with RegionRunConfig class and it works fine except for the YearType which is an Enum.

```
from glambie.config.config_classes import GlambieRunConfig, RegionRunConfig
import yaml


def main():
    # create a list of custom classes, some of which contain nested classes
    # config_path = "tests/test_data/configs/config_pilot_study.yaml"
    config_path = "tests/test_data/configs/test_config.yaml"
    stuff = [GlambieRunConfig.from_yaml(config_path).regions[0]]

    # use a custom representer to forcibly convert the object to its own dictionary
    def region_run_config_class_representer(yaml_dumper, object_to_represent):
        return yaml_dumper.represent_dict(object_to_represent.__dict__)

    yaml.add_representer(RegionRunConfig, region_run_config_class_representer)

    print(yaml.dump(stuff))


if __name__ == '__main__':
    main()
```

This gives the result:

```
- region_name: iceland
  region_run_settings:
    altimetry:
      exclude_annual_datasets:
      - xxxxxxxxxxxxxxxxxxxx
      exclude_trend_datasets:
      - xxxxxxxxxxxxxxxxxxxx
    demdiff_and_glaciological:
      exclude_annual_datasets:
      - sharks
      exclude_trend_datasets:
      - hugonnet
    gravimetry:
      exclude_annual_datasets:
      - ciraci
      exclude_trend_datasets:
      - wouters
  seasonal_correction_dataset:
    data_group: glaciological
    user_group: wgms_sine
  year_type: !!python/object/apply:glambie.const.constants.YearType
  - glaciological
```

I changed the __repr__() of the YearType class to be self.value (which in this case would be "glaciological"), however that only changes when printing the dict, not when passing into yaml dump.

I can only think of hacky ways to solve this or making custom to_dict() methods for all classes?

Reply:
So, dumping to YAML is effectively a form of serialisation. Serialisation is not the same problem as "make this human readable", because serialisation needs to be reversible (i.e. also machine readable). `repr()` is designed purely for human readability, not serialisation, and so doesn't really have anything to do with saving to file (other than possibly for logging).

Given that you're trying to strip out type indicators, I'm going to assume that you don't want to make this a reversible operation. In that case, it would make sense to simply write additional class representers for anything that doesn't work using the above example. That means no, you do not need to write custom to_dict() methods for all classes (or indeed any classes), and it's important to understand that (a) a yaml representer is not a means of converting something to a dictionary and (b) you don't need to convert something to a dictionary in order to save it sensibly to YAML.

Reply:
Ah got it now. Thanks!

Reply:
Madde a class representer for YearType and RegionRunConfig in yaml_helpers.py